File: hackit/wireless/plugins/ruby/scripts/airodump_ng.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'time'
require 'open3'

def emit(event, iface, bssid, ssid, channel, data)
  puts JSON.generate(event: event, iface: iface, bssid: bssid, ssid: ssid, channel: channel, data: data, timestamp: Time.now.iso8601)
  $stdout.flush
end

def parse_csv_line(line, sep = ',')
  fields = []
  current = +''
  in_quotes = false
  line.each_char do |c|
    case c
    when '"' then in_quotes = !in_quotes
    when sep then (fields << current; current = +'') unless in_quotes
    else current << c
    end
  end
  fields << current
  fields.map(&:strip)
end

def parse_airodump(prefix)
  csv_path = "#{prefix}-01.csv"
  return unless File.exist?(csv_path)

  aps = {}
  clients = {}
  section = :ap

  File.readlines(csv_path).each do |line|
    next if line.strip.empty?
    line = line.encode('UTF-8', invalid: :replace, undef: :replace)
    if line.start_with?('BSSID,')
      section = :ap
      next
    elsif line.start_with?('Station MAC,')
      section = :client
      next
    end

    fields = parse_csv_line(line)
    if section == :ap && fields.size >= 14
      bssid = fields[0].upcase
      next if bssid.empty?
      aps[bssid] = {
        first_seen: fields[1], last_seen: fields[2], channel: fields[3].to_i,
        speed: fields[4], privacy: fields[5], cipher: fields[6],
        auth: fields[7], power: fields[8].to_i, beacon: fields[9].to_i,
        llc: fields[10].to_i, data_packets: fields[11].to_i, rate: fields[12],
        essid: fields[13]
      }
    elsif section == :client && fields.size >= 7
      mac = fields[0].upcase
      bssid = fields[5].upcase
      clients[mac] = {
        first_seen: fields[1], last_seen: fields[2], power: fields[3].to_i,
        packets: fields[4].to_i, bssid: bssid, rate: fields[6]
      }
    end
  end

  { aps: aps, clients: clients }
end

begin
  iface = ARGV[0]
  output_prefix = ARGV[1] || 'dump'
  channel = (ARGV[2] || '0').to_i

  raise 'interface required' unless iface

  emit('airodump_start', iface, '', '', channel, { output_prefix: output_prefix })
  airodump_bin = `which airodump-ng 2>/dev/null`.strip
  raise 'airodump-ng not found' if airodump_bin.empty?

  cmd = [airodump_bin, iface, '--write', output_prefix, '--write-interval', '1']
  cmd += ['--channel', channel.to_s] if channel > 0

  emit('airodump_launch', iface, '', '', channel, { command: cmd.join(' ') })

  Open3.popen3(*cmd) do |stdin, stdout, stderr, thread|
    stdin.close
    [stdout, stderr].each(&:close)

    loop do
      sleep 2
      break unless thread.alive?

      data = parse_airodump(output_prefix)
      data[:aps].each do |bssid, info|
        emit('ap_found', iface, bssid, info[:essid], info[:channel], info)
      end
      data[:clients].each do |mac, info|
        emit('client_found', iface, mac, '', 0, info)
      end
    end

    thread.join
  end

  emit('airodump_complete', iface, '', '', channel, { output_prefix: output_prefix })

rescue Interrupt
  emit('airodump_interrupted', iface || '', '', '', 0, {})
rescue => e
  $stderr.puts JSON.generate(event: 'error', data: { message: e.message, backtrace: e.backtrace&.first }, timestamp: Time.now.iso8601)
  exit 1
end

File: hackit/wireless/plugins/lua/scripts/airodump_ng.lua
local iface = arg[1]
local output_prefix = arg[2] or "/tmp/airodump_" .. iface
local channel = arg[3] or ""

local function json_escape(s)
  return '"' .. tostring(s):gsub('\\', '\\\\'):gsub('"', '\\"'):gsub('\n', '\\n'):gsub('\r', '\\r') .. '"'
end

local function telemetry(event, data)
  local ts = tostring(os.time())
  io.write('{"event":' .. json_escape(event) .. ',"iface":' .. json_escape(iface) .. ',"bssid":"","ssid":"","channel":' .. json_escape(channel) .. ',"data":' .. data .. ',"timestamp":' .. json_escape(ts) .. '}\n')
  io.flush()
end

local function check_monitor_mode()
  local f = io.popen("iw dev " .. iface .. " info 2>/dev/null")
  if not f then return false end
  local out = f:read("*a")
  f:close()
  if out:find("type monitor") then return true end
  telemetry("monitor_fail", '{"error":"Interface not in monitor mode"}')
  return false
end

local function parse_csv_line(line)
  local parts = {}
  for part in line:gmatch('([^,]+)') do
    table.insert(parts, part:match('^%s*(.-)%s*$'))
  end
  return parts
end

local function run_airodump()
  local cmd = "airodump-ng --write " .. output_prefix .. " --output-format csv,pcap " .. iface
  if channel and #channel > 0 then
    cmd = cmd .. " --channel " .. channel
  end
  cmd = cmd .. " 2>/dev/null &"
  telemetry("airodump_start", '{"cmd":' .. json_escape(cmd) .. ',"output":' .. json_escape(output_prefix) .. '}')
  local pid = io.popen("echo $!", "r")
  os.execute(cmd .. " &")
  if pid then
    local p = pid:read("*a")
    pid:close()
    telemetry("airodump_pid", '{"pid":' .. json_escape(p:match("%d+") or "unknown") .. '}')
  end
end

local function parse_csv()
  local csv_file = output_prefix .. "-01.csv"
  local seen_aps = {}
  local seen_clients = {}
  local in_clients = false
  for i = 1, 30 do
    os.execute("sleep 2")
    local f = io.open(csv_file, "r")
    if not f then
      telemetry("csv_wait", '{"attempt":' .. tostring(i) .. ',"file":' .. json_escape(csv_file) .. '}')
      goto continue
    end
    local ap_section = true
    for line in f:lines() do
      if line:find("^BSSID,") or line:find("^Station MAC,") then
        if line:find("^Station MAC,") then ap_section = false end
        goto nextline
      end
      if #line < 10 then goto nextline end
      local parts = parse_csv_line(line)
      if #parts < 6 then goto nextline end
      if ap_section then
        local b = parts[1]
        if b and b:match("^[%x:]+$") and #b == 17 then
          local pwr = parts[3] or "0"
          local ch = parts[5] or "0"
          local enc = (parts[6] or ""):gsub('"', '')
          local essid = (parts[13] or ""):gsub('"', '')
          if not seen_aps[b] then
            seen_aps[b] = true
            telemetry("ap_found", '{"bssid":' .. json_escape(b) .. ',"power":' .. json_escape(pwr) .. ',"channel":' .. json_escape(ch) .. ',"encryption":' .. json_escape(enc) .. ',"ssid":' .. json_escape(essid) .. '}')
          end
        end
      else
        local mac = parts[0] or parts[1]
        if mac and mac:match("^[%x:]+$") and #mac == 17 then
          local bssid_col = parts[client_bssid_col or 5] or ""
          local pwr = parts[3] or "0"
          if not seen_clients[mac] then
            seen_clients[mac] = true
            telemetry("client_found", '{"mac":' .. json_escape(mac) .. ',"bssid":' .. json_escape(bssid_col) .. ',"power":' .. json_escape(pwr) .. '}')
          end
        end
      end
      ::nextline::
    end
    f:close()
    telemetry("csv_parse", '{"aps_found":' .. tostring(#seen_aps) .. ',"clients_found":' .. tostring(#seen_clients) .. '}')
    ::continue::
  end
end

local function monitor_pmkid_handshakes()
  local pcap_file = output_prefix .. "-01.cap"
  for i = 1, 30 do
    os.execute("sleep 5")
    local f = io.popen("tshark -r " .. pcap_file .. " -Y 'eapol' -T fields -e wlan.sa -e wlan.da -e eapol.keydes.info 2>/dev/null | sort -u | head -20")
    if f then
      local out = f:read("*a")
      f:close()
      if out and #out > 0 then
        for line in out:gmatch("[^\n]+") do
          local sa, da, kinfo = line:match("([%x:]+)%s+([%x:]+)%s+(.+)")
          if sa and da then
            telemetry("eapol_capture", '{"source":' .. json_escape(sa) .. ',"dest":' .. json_escape(da) .. ',"key_info":' .. json_escape(kinfo or "") .. '}')
          end
        end
      end
    end
    local h = io.popen("hcxpcapngtool -o /dev/stdout " .. pcap_file .. " 2>/dev/null | head -5")
    if h then
      local hout = h:read("*a")
      h:close()
      if hout and #hout > 0 then
        telemetry("pmkid_found", '{"hash":' .. json_escape(hout:sub(1, 200)) .. '}')
      end
    end
  end
end

if not iface then
  telemetry("error", '{"error":"Usage: airodump_ng.lua <interface> [output_prefix] [channel]"}')
  os.exit(1)
end

if not check_monitor_mode() then
  telemetry("error", '{"error":"Monitor mode required on ' .. iface .. '"}')
  os.exit(1)
end

local ok, err = pcall(function()
  run_airodump()
  parse_csv()
  monitor_pmkid_handshakes()
end)

if not ok then
  telemetry("error", '{"error":' .. json_escape(tostring(err)) .. '}')
  os.exit(1)
end

telemetry("complete", '{"output_prefix":' .. json_escape(output_prefix) .. '}')
